File: perl/psl.pl
use strict;
use warnings;
use feature qw(say);
use Domain::PublicSuffix;

my $suffix = Domain::PublicSuffix->new({data_file => '/tmp/public_suffix_list.dat'});
for (qw(www.google.co.jp foo.bar.yokohama.jp bar.yokohama.jp yokohama.jp jp foo.bar.platform.sh bar.platform.sh platform.sh sh foo.s3.amazonaws.com s3.amazonaws.com foo.s1.amazonaws.com s1.amazonaws.com)) {
    my $root = $suffix->get_root_domain($_);
    say "[$_]";
    say "root: ", ($root // '(undef)');
    say "public suffix: ", ($suffix->suffix // '(undef)');
    say "tld: ", ($suffix->tld // '(undef)');
    say "error: ", ($suffix->error // '(undef)');
    say "";
}

File: python/psl.py
from publicsuffix2 import PublicSuffixList

psl = PublicSuffixList(psl_file="/tmp/public_suffix_list.dat")
for domain in [
    "www.google.co.jp",
    "foo.bar.yokohama.jp",
    "bar.yokohama.jp",
    "yokohama.jp",
    "jp",
    "foo.bar.platform.sh",
    "bar.platform.sh",
    "platform.sh",
    "sh",
    "foo.s3.amazonaws.com",
    "s3.amazonaws.com",
    "foo.s1.amazonaws.com",
    "s1.amazonaws.com",
]:
    print(f"[{domain}]")
    print(f"root: {psl.get_public_suffix(domain)}")
    print(f"public suffix: {psl.get_tld(domain)}")
    print()
